src/dbnode/persist/fs/write.go

```diff
@@ -327,13 +329,39 @@ func (w *writer) Close() error {
 	}
 	// NB(xichen): only write out the checkpoint file if there are no errors
 	// encountered between calling writer.Open() and writer.Close().
-	if err := w.writeCheckpointFile(); err != nil {
+	if err := writeCheckpointFile(
+		w.checkpointFilePath,
+		w.digestFdWithDigestContents.Digest().Sum32(),
+		w.digestBuf,
+		w.newFileMode,
+	); err != nil {
 		w.err = err
 		return err
 	}
 	return nil
 }
 
+func (w *writer) DeferClose() (persist.DataCloser, error) {
+	err := w.close()
+	if w.err != nil {
+		return nil, w.err
+	}
+	if err != nil {
+		w.err = err
+		return nil, err
+	}
+	checkpointFilePath := w.checkpointFilePath
+	digestChecksum := w.digestFdWithDigestContents.Digest().Sum32()
+	return func() error {
+		return writeCheckpointFile(
+			checkpointFilePath,
+			digestChecksum,
+			digest.NewBuffer(),
+			w.newFileMode,
```

Comment:
Maybe also take ref to `w.newFileMode` above declaring the lambda? If for whatever reason that changed between reuse?